File: python/benchmark/core/benchmark_runner.py
from __future__ import annotations

import sys
import threading
import time
from concurrent.futures import ThreadPoolExecutor, wait
from typing import List, Optional

from benchmark.core._sync import CountDownLatch
from benchmark.core.benchmark_config import BenchmarkConfig
from benchmark.core.benchmark_result import BenchmarkResult
from benchmark.core.benchmark_task import BenchmarkTask
from benchmark.metrics.metrics_collector import MetricsCollector


class BenchmarkRunner:
    """Drives a BenchmarkTask across all configured concurrency levels."""

    def __init__(self, config: BenchmarkConfig) -> None:
        self._config = config

    def run(self, task: BenchmarkTask) -> List[BenchmarkResult]:
        results: List[BenchmarkResult] = []
        print()
        print("+======================================================+")
        print(f"|  Benchmarking: {task.name():<37}|")
        print("+======================================================+")

        task.setup()

        for concurrency in self._config.concurrency_levels:
            print(f"\n  >>  concurrency = {concurrency}")
            result = self._run_level(task, concurrency)
            results.append(result)
            print(f"     {result}")
            time.sleep(0.5)

        task.teardown()
        return results

    def _run_level(self, task: BenchmarkTask, concurrency: int) -> BenchmarkResult:
        self._run_workers(
            task, concurrency, task.warmup_iterations(), None, None, None, None
        )

        iteration_times_ns: List[int] = []
        times_lock = threading.Lock()
        total_iterations = _AtomicInt()
        error_count = _AtomicInt()

        collector = MetricsCollector(self._config.metrics_sample_interval_ms)
        wall_start = time.time() * 1000.0
        collector.start()

        self._run_workers(
            task,
            concurrency,
            task.iterations_per_worker(),
            iteration_times_ns,
            times_lock,
            total_iterations,
            error_count,
        )

        snapshots = collector.stop()
        collector.shutdown()
        wall_end = time.time() * 1000.0
        total_duration_ms = int(wall_end - wall_start)

        if error_count.value > 0:
            print(
                f"     WARNING: {error_count.value} iteration(s) threw exceptions (included in timing)"
            )

        return BenchmarkResult(
            task_name=task.name(),
            concurrency_level=concurrency,
            total_iterations=total_iterations.value,
            total_duration_ms=total_duration_ms,
            iteration_times_ns=tuple(iteration_times_ns),
            metrics_snapshots=snapshots,
            avg_cpu_percent=MetricsCollector.average_cpu(snapshots),
            peak_cpu_percent=MetricsCollector.peak_cpu(snapshots),
            avg_memory_mb=MetricsCollector.average_memory_mb(snapshots),
            peak_memory_mb=MetricsCollector.peak_memory_mb(snapshots),
        )

    def _run_workers(
        self,
        task: BenchmarkTask,
        concurrency: int,
        iterations: int,
        iteration_times_ns: Optional[List[int]],
        times_lock: Optional[threading.Lock],
        total_iterations: Optional["_AtomicInt"],
        error_count: Optional["_AtomicInt"],
    ) -> None:
        ready = CountDownLatch(concurrency)
        start = CountDownLatch(1)

        def worker(worker_id: int) -> None:
            ready.count_down()
            start.await_zero()
            for _ in range(iterations):
                t0 = time.perf_counter_ns()
                try:
                    task.execute(worker_id)
                except Exception as ex:
                    if error_count is not None:
                        error_count.add(1)
                    if self._config.verbose:
                        print(
                            f"  Worker {worker_id} failed: {ex}",
                            file=sys.stderr,
                        )
                elapsed = time.perf_counter_ns() - t0
                if iteration_times_ns is not None and times_lock is not None:
                    with times_lock:
                        iteration_times_ns.append(elapsed)
                if total_iterations is not None:
                    total_iterations.add(1)

        with ThreadPoolExecutor(max_workers=concurrency) as executor:
            futures = [executor.submit(worker, w) for w in range(concurrency)]
            ready.await_zero()
            start.count_down()
            done, not_done = wait(futures, timeout=self._config.task_timeout_seconds)
            if not_done:
                print(
                    f"  WARNING: Concurrency level {concurrency} timed out after "
                    f"{self._config.task_timeout_seconds}s"
                )
                for f in not_done:
                    f.cancel()


class _AtomicInt:
    def __init__(self) -> None:
        self._v = 0
        self._lock = threading.Lock()

    def add(self, n: int = 1) -> None:
        with self._lock:
            self._v += n

    @property
    def value(self) -> int:
        with self._lock:
            return self._v

File: benchmark-framework-source/src/main/java/com/benchmark/core/BenchmarkRunner.java
package com.benchmark.core;

import com.benchmark.metrics.MetricsCollector;
import com.benchmark.metrics.MetricsSnapshot;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Drives a {@link BenchmarkTask} across all configured concurrency levels.
 *
 * <p>For each level the runner:
 * <ol>
 *   <li>Runs warm-up iterations (discarded)</li>
 *   <li>Starts the metrics-collection background thread</li>
 *   <li>Launches N worker threads, each executing the task repeatedly</li>
 *   <li>Waits for all workers to finish (or times out)</li>
 *   <li>Stops the metrics collector</li>
 *   <li>Assembles a {@link BenchmarkResult}</li>
 * </ol>
 * </p>
 */
public final class BenchmarkRunner {

    private final BenchmarkConfig config;

    public BenchmarkRunner(BenchmarkConfig config) {
        this.config = config;
    }

    /**
     * Run a single task across all concurrency levels.
     *
     * @return one result per concurrency level
     */
    public List<BenchmarkResult> run(BenchmarkTask task) throws Exception {
        List<BenchmarkResult> results = new ArrayList<>();

        System.out.printf("%n╔══════════════════════════════════════════════════════╗%n");
        System.out.printf("║  Benchmarking: %-37s║%n", task.name());
        System.out.printf("╚══════════════════════════════════════════════════════╝%n");

        task.setup();

        for (int concurrency : config.getConcurrencyLevels()) {
            System.out.printf("%n  ▶  concurrency = %d%n", concurrency);
            BenchmarkResult result = runLevel(task, concurrency);
            results.add(result);
            System.out.println("     " + result);

            // Brief pause between levels to let the JVM settle
            Thread.sleep(500);
        }

        task.teardown();
        return results;
    }

    // -------------------------------------------------------------------------

    private BenchmarkResult runLevel(BenchmarkTask task, int concurrency) throws Exception {

        // --- warm-up phase (not measured) ---
        runWorkers(task, concurrency, task.warmupIterations(), null, null, null);

        // --- measured phase ---
        List<Long> iterationTimesNs     = new CopyOnWriteArrayList<>();
        AtomicLong totalIterations      = new AtomicLong();
        AtomicInteger errorCount        = new AtomicInteger();
        MetricsCollector metricsCollector = new MetricsCollector(config.getMetricsSampleIntervalMs());

        long wallStart = System.currentTimeMillis();
        metricsCollector.start();

        runWorkers(task, concurrency, task.iterationsPerWorker(),
                   iterationTimesNs, totalIterations, errorCount);

        List<MetricsSnapshot> snapshots = metricsCollector.stop();
        long wallEnd = System.currentTimeMillis();
        metricsCollector.shutdown();

        long totalDurationMs = wallEnd - wallStart;

        if (errorCount.get() > 0) {
            System.out.printf("     ⚠  %d iteration(s) threw exceptions (included in timing)%n",
                              errorCount.get());
        }

        return new BenchmarkResult(
                task.name(),
                concurrency,
                totalIterations.get(),
                totalDurationMs,
                iterationTimesNs,
                snapshots,
                MetricsCollector.averageCpu(snapshots),
                MetricsCollector.peakCpu(snapshots),
                MetricsCollector.averageMemoryMB(snapshots),
                MetricsCollector.peakMemoryMB(snapshots)
        );
    }

    /**
     * Spin up {@code concurrency} threads, each running {@code iterations} times.
     * If {@code timings} / {@code counter} are null this is a warm-up run.
     */
    private void runWorkers(BenchmarkTask task,
                            int concurrency,
                            int iterations,
                            List<Long> timings,
                            AtomicLong counter,
                            AtomicInteger errors) throws Exception {

        ExecutorService pool = Executors.newFixedThreadPool(concurrency);
        CountDownLatch ready  = new CountDownLatch(concurrency);
        CountDownLatch start  = new CountDownLatch(1);
        List<Future<?>> futures = new ArrayList<>();

        for (int w = 0; w < concurrency; w++) {
            final int workerId = w;
            futures.add(pool.submit(() -> {
                ready.countDown();
                try { start.await(); } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    return;
                }
                for (int i = 0; i < iterations; i++) {
                    long t0 = System.nanoTime();
                    try {
                        task.execute(workerId);
                    } catch (Exception ex) {
                        if (errors != null) errors.incrementAndGet();
                        if (config.isVerbose())
                            System.err.printf("  Worker %d iteration %d failed: %s%n",
                                              workerId, i, ex.getMessage());
                    }
                    long elapsed = System.nanoTime() - t0;
                    if (timings  != null) timings.add(elapsed);
                    if (counter  != null) counter.incrementAndGet();
                }
            }));
        }

        // Wait for all workers to be ready, then release them simultaneously
        ready.await();
        start.countDown();

        pool.shutdown();
        boolean finished = pool.awaitTermination(config.getTaskTimeoutSeconds(), TimeUnit.SECONDS);
        if (!finished) {
            pool.shutdownNow();
            System.out.printf("  ⚠  Concurrency level %d timed out after %ds%n",
                              concurrency, config.getTaskTimeoutSeconds());
        }
    }
}
